### PR TITLE
Cross join to inner rewrite

### LINKED ISSUE #19856: AND (cond OR cond) prevent CROSS JOIN to INNER JOIN rewrite
**Describe the bug**
It's possible to rewrite CROSS JOIN TO INNER JOIN in certain conditions, but clickhouse doesn't do that when we have ```AND (cond OR cond)``` in WHERE clause.

**Does it reproduce on recent release?**
Yes

**How to reproduce**
Clickhouse version 21.2

```
EXPLAIN SYNTAX
WITH tbl AS
    (
        SELECT *
        FROM numbers(1000)
    )
SELECT count(x.number)
FROM tbl AS x
, tbl AS z
WHERE x.number = z.number

Query id: ea50c6a4-618d-4f87-a089-01ef2ba956fe

┌─explain──────────────────────────────────────┐
│ WITH tbl AS                                  │
│     (                                        │
│         SELECT *                             │
│         FROM numbers(1000)                   │
│     )                                        │
│ SELECT count(number)                         │
│ FROM tbl AS x                                │
│ ALL INNER JOIN tbl AS z ON number = z.number │
│ WHERE number = z.number                      │
└──────────────────────────────────────────────┘


EXPLAIN SYNTAX
WITH tbl AS
    (
        SELECT *
        FROM numbers(1000)
    )
SELECT count(x.number)
FROM tbl AS x
, tbl AS z
WHERE (x.number = z.number) AND ((x.number = 1) OR (x.number = 2))

Query id: 4f3f6e5a-f7fc-41c0-a2f2-675525e5d63e

┌─explain──────────────────────────────────────────────────────┐
│ WITH tbl AS                                                  │
│     (                                                        │
│         SELECT *                                             │
│         FROM numbers(1000)                                   │
│     )                                                        │
│ SELECT count(number)                                         │
│ FROM tbl AS x                                                │
│ CROSS JOIN tbl AS z                                          │
│ WHERE (number = z.number) AND ((number = 1) OR (number = 2)) │
└──────────────────────────────────────────────────────────────┘


EXPLAIN SYNTAX
WITH tbl AS
    (
        SELECT *
        FROM numbers(1000)
    )
SELECT count(x.number)
FROM tbl AS x
, tbl AS z
WHERE (x.number = z.number) AND 1

Query id: 3ba0c863-8432-441a-b04c-e06047436cbc

┌─explain─────────────────────────┐
│ WITH tbl AS                     │
│     (                           │
│         SELECT *                │
│         FROM numbers(1000)      │
│     )                           │
│ SELECT count(number)            │
│ FROM tbl AS x                   │
│ CROSS JOIN tbl AS z             │
│ WHERE (number = z.number) AND 1 │
└─────────────────────────────────┘

```

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

- Support more cases to rewrite `CROSS JOIN` to `INNER JOIN`.

Close #19856

### CODE REVIEW

**Comment by vdimir** [comment]
For now rules fort `==` operator in `JOIN ON` more strict than in `WHERE` section. `JOIN ON` requires exact type match, but `WHERE` allows implicit type conversion. So, some queries may do not work after rewrite. But after changes from PR #19885 would be merged we can move expression safely.

**Comment by akuzm** [comment]
By the way, why do we even have cross/comma join as separate concepts? They are fully equivalent to a clauseless `INNER` join. There is some difference in operator precedence with comma, but it shouldn't survive past the parsing stage.

Probably the reason is that we don't have a suitable structure for representing the join tree, and use raw AST for this?

File: src/Interpreters/CrossToInnerJoinVisitor.cpp [hunk 1]
```diff
@@ -80,147 +81,150 @@ struct JoinedElement
...
+    return members;
+}
 
-    size_t canMoveEqualsToJoinOn(const ASTFunction & node)
-    {
-        if (!node.arguments)
-            throw Exception("Logical error: function requires arguments", ErrorCodes::LOGICAL_ERROR);
-        if (node.arguments->children.size() != 2)
-            return false;
+std::optional<size_t> getIdentMembership(const ASTIdentifier & ident, const std::vector<TableWithColumnNamesAndTypes> & tables)
```
Comment (akuzm) [inline src/Interpreters/CrossToInnerJoinVisitor.cpp:134]: Nice, this almost starts to look like the proper join clause infrastructure. Eventually we should have an explicit data structure describing the join tree, and then for each WHERE clause, check which relations participate in it, and put it to the appropriate level in this tree.

**Review by akuzm — APPROVED** [review]

File: src/Interpreters/CrossToInnerJoinVisitor.cpp (hunk 1 above)
Comment (vdimir) [reply src/Interpreters/CrossToInnerJoinVisitor.cpp:134]: I also tried to use this code in https://github.com/ClickHouse/ClickHouse/pull/21370 to fix problem with `Merge` engine. While executing JOIN with `Merge` there are step when we remove right table from query and perform analysis on it (I don't completly figure out why it is needed, maybe it requires other approach). But expression that uses right table left in query and it leads to error. This code can be used to remove them too.

**Pull request merged**